### PR TITLE
CLOUDP-278492: Avoid panic when changing type of existing cluster

### PR DESCRIPTION
**Pull request by helderjs**
### All Submissions:

* [x] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if there is one).
* [ ] Update docs/release-notes/release-notes-template.md if your changes should be included in the release notes for the next release.

### CODE REVIEW

File: pkg/controller/atlasdeployment/atlasdeployment_controller.go [hunk 1]
```diff
@@ -162,6 +162,10 @@ func (r *AtlasDeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 	wasDeleted := !atlasDeployment.GetDeletionTimestamp().IsZero()
 	existsInAtlas := deploymentInAtlas != nil
 
+	if existsInAtlas && fmt.Sprintf("%T", deploymentInAKO) != fmt.Sprintf("%T", deploymentInAtlas) {
```
Comment (s-urbaniak) [inline pkg/controller/atlasdeployment/atlasdeployment_controller.go:165]: No, please don't switch on string representations of types, this is not right. A low level `%T` language type primitive must not be used to assert high level business logic. You can accomplish this using different techniques.

I.e. you can extend the `translation.Deployment` interface with a `IsServerless()` method. In the case of `NewDeployment` it returns it based on the spec and in the case of the `GetDeployment` service it returns it based whether an underlying serverless or regular deployment was found.

Comment (s-urbaniak) [reply pkg/controller/atlasdeployment/atlasdeployment_controller.go:165]: the assertion would then change to something like:
```
	if existsInAtlas && deploymentInAKO.IsServerless() != deploymentInAtlas.IsServerless() {
```

File: pkg/controller/atlasdeployment/atlasdeployment_controller_test.go [hunk 2]
```diff
@@ -1422,3 +1424,175 @@ func TestGetProjectFromKube(t *testing.T) {
 		})
 	}
 }
+
+func TestChangeDeploymentType(t *testing.T) {
+	tests := map[string]struct {
+		deployment *akov2.AtlasDeployment
+	}{
+		"should fail when existing cluster is regular but manifest defines a serverless instance": {
```
Comment (josvazg) [inline pkg/controller/atlasdeployment/atlasdeployment_controller_test.go:1432]: I was expecting two inputs; existing and new spec. How can this work with one input only?

Comment (helderjs) [reply pkg/controller/atlasdeployment/atlasdeployment_controller_test.go:1432]: The input is an already-created deployment that will conflict with its mocked counterpart (lines 1534-1560) coming from Atlas.

**Review by s-urbaniak — APPROVED** [review]

File: pkg/controller/atlasdeployment/atlasdeployment_controller_test.go (hunk 2 above)
Comment (josvazg) [reply pkg/controller/atlasdeployment/atlasdeployment_controller_test.go:1432]: Fine. I'd rather make the test logic dumb simple and ensure the inputs are more explicit, but will not block the PR on this.

**Review by josvazg — APPROVED** [review]
LGTM, providing CI agrees

**Pull request merged**